### PR TITLE
Implement Shadow Impostors

### PR DESCRIPTION
**Pull request by TokisanGames**
This PR implements shadow impostors by adding a new variable called `max_shadow_caster`. This allows us to increase performance by rendering shadows based on lower resolution objects. I'm currently using this in my game on trees and buildings.

Shadows won't use a LOD higher than the one specified. If you play the test project and look at the new shadow caster objects, you'll see LOD0 is using a lower LOD shadow.

shadow_caster_test

### CODE REVIEW

File: gdnative/multi-lod/src/lod.cpp [hunk 1]
```diff
@@ -358,9 +292,43 @@ void LOD::update_lod_multipliers_from_manager() {
...
-    show ? lod_object->show() : lod_object->hide();
-    if (disable_processing) {
-        call_deferred("set_node_processing", lod_object, show);
+void LOD::show_lod(int lod) {
+    // This function is safe to send lods >= LOD_COUNT, which results in every LOD hidden.
+    // What is caught and returned in this first conditional, resulting in nothing are
+    // if we're already on this LOD level,
+    // or LOD < LOD_COUNT and it doesn't exist.
+    // This means last LOD will remain visible until it hits actual_hide_distance.
+    if (lod == current_lod || (lod < LOD_COUNT && ! lods[lod])) {
```
Comment (puchik) [inline gdnative/multi-lod/src/lod.cpp]: Can you please wrap (lod  < LOD_COUNT) in parentheses and remove the space after the exclamation mark i.e. !lods[lod]

File: gdnative/multi-lod/src/lod.cpp [hunk 2]
```diff
@@ -358,9 +292,43 @@ void LOD::update_lod_multipliers_from_manager() {
...
+        return;
+    }
+
+    // Count backwards to hit shadow caster first
+    for(int i = LOD_COUNT - 1; i >= 0 ; i--) {
+        if (lods[i]) {
```
Comment (puchik) [inline gdnative/multi-lod/src/lod.cpp:320]: You could put these two if statements on the same line/combine them.

File: gdnative/multi-lod/src/lod.cpp [hunk 3]
```diff
@@ -358,9 +292,43 @@ void LOD::update_lod_multipliers_from_manager() {
     }
 }
 
-void LOD::show_lod(Spatial* lod_object, bool show) {
-    show ? lod_object->show() : lod_object->hide();
-    if (disable_processing) {
-        call_deferred("set_node_processing", lod_object, show);
+void LOD::show_lod(int lod) {
+    // This function is safe to send lods >= LOD_COUNT, which results in every LOD hidden.
+    // What is caught and returned in this first conditional, resulting in nothing are
```
Comment (puchik) [inline gdnative/multi-lod/src/lod.cpp]: Can you reword this comment? It was a bit difficult to process at first. Something like (if I understood it correctly) "It is safe to pass lod >= LOD_Count to this function, which will result in every LOD being hidden. The first conditional checks whether we are already on this LOD level or if the requested LOD doesn't exist. This means the last LOD will remain visible until it hits actual_hide_distance."

**Comment by puchik** [comment]
This looks good. I haven't tested it yet but the draft is quite nice (appreciate the GIF). I like the refactors (much cleaner show/hide logic) and addition of more Godot-specific calls. Left a few nitpicks as review for now.

**Comment by puchik** [comment]
Forgot to hit the submit review button 😄 You should see it now.

File: gdnative/multi-lod/src/lod.cpp (hunk 1 above)
Comment (TokisanGames) [reply gdnative/multi-lod/src/lod.cpp]: Done.

File: gdnative/multi-lod/src/lod.cpp (hunk 2 above)
Comment (TokisanGames) [reply gdnative/multi-lod/src/lod.cpp:320]: L307 is separate because it also validates lods[i] for L327-329, disable processing. 

Incidentally, I think disable processing needs to be adjusted in another PR. Currently it enables/disables _process and _physics_process on the LOD meshes themselves. That's not what I need, and I suspect others don't need that either.

Consider a scene structure like this:
```
PhysicsBody (static/kinematic + Enemy.gd)
\
 - ImmediateGeometry (MTLOD)
  \
    - LOD0 (MeshInstance)
    - LOD1 (MeshInstance)
    - LOD2 (MeshInstance)
    - LOD3 (MeshInstance)
 - CollsionShape
 - Other nodes like a hitbox, animation player, sound player, etc
```

The meshinstances aren't going to be running scripts individually, so don't need them to enable or disable processing. I need the parent node, the physicsbody to disable or enable processing. 

And I need to do more than just set physics_process and _process. I want to disable a collisionshape, a statemachine, an animation player, and a particle emitter and more. I can do all of these for visibility by customizing `show()` and `hide()`, but not with this disable process. 

So I think this should change to provide a NodePath, then call a user defined function, perhaps a funcref or a String, and/or a default `enable()` and `disable()`, or it could emit a signal for the node to capture and do it's own disabling and enabling.

I currently have written this in a gdscript ProximityManager that disables objects that are outside of player proximity, but it's a bit redundant with and less efficient than MTLOD.

What do you think?

Comment (TokisanGames) [reply gdnative/multi-lod/src/lod.cpp:320]: Also, in the old version and new, neither `enabled`, nor `disable_processing` work for me. When `Enabled` is turned off, LODs are still processed. When `disable_processing is checked, I get many errors about it. 

```
ERROR: Error calling deferred method: 'ImmediateGeometry(lod.gdns)::set_node_processing': Method not found..
   at: _call_function (core/message_queue.cpp:243)
```

**Comment by TokisanGames** [comment]
Your old show_lod functionality showed the highest active LOD until hide_distance. That was my intention as well, but my code didn't do that. If LOD3 didn't exist, LOD2 would be hidden between lod3_distance and hide_distance. This commit fixes that and makes the changes you requested, and another adds documentation.

File: gdnative/multi-lod/src/lod.cpp (hunk 3 above)
Comment (TokisanGames) [reply gdnative/multi-lod/src/lod.cpp]: done.

File: gdnative/multi-lod/src/lod.cpp (hunk 2 above)
Comment (puchik) [reply gdnative/multi-lod/src/lod.cpp:320]: Aha I see what you mean now. As in, `process` can be toggled as long as the node exists and it doesn't necessarily need to be in the tree?

* RE where `process` is toggled

The intention was, since toggling process propagates through the children (does it not?), it would disable scripts that, for example, might make a camera follow a player around etc or maybe disabling some sort of script-based effect. 

Additionally, the main LOD isn't just for meshes/`VisualInstance`s (if you're not using screen-percentage calculation), so you could use it to disable such objects on their own even without any mesh.

Do you mean that you have a script that can disable the entire object and not just the `process`? I would be glad to add/merge that, since iirc at the time of my implementation either there was no way to do this or I was not able to find it in the documentation and that was the my intention (which ended up being "close enough").

* RE triggering other events on various LOD levels:

Adding a signal to be called when `process` is toggled or when a particular LOD level is selected or the LOD is hidden could be a valuable addition. I would say you use the Godot Signal system to accomplish this and make it optional/a setting which is defaulted to off to avoid overhead (though it could be useful to add a way to notify users of this in case they are stuck trying to use the signals, only to realize the setting to send it was disabled).

* RE errors when enabling toggle process:

Hmm... it looks like `register_method` is not called on `set_node_processing` in [`_register_methods`](https://github.com/puchik/godot-extras/blob/57e595bfc0a8d787615c03369f1e116bdbbe484d/gdnative/multi-lod/src/lod.cpp#L7). Very weird, because I remember testing and working on this one very thoroughly since there were issues with too many deferred calls. It's possible that maybe the version of Godot I used while developing handled deferred calls on its own object differently (since it would be called on this class, and not `ImmediateGeometry`). Adding the `register_method` call _should_ fix that.

* RE `enabled`:

I think this is somewhat misleading or should be hidden. Looks like it is actually [controlled by itself](https://github.com/puchik/godot-extras/blob/57e595bfc0a8d787615c03369f1e116bdbbe484d/gdnative/multi-lod/src/lod.cpp#L70) to determine whether an object is active internally, like [here](https://github.com/puchik/godot-extras/blob/57e595bfc0a8d787615c03369f1e116bdbbe484d/gdnative/multi-lod/src/lod.cpp#L266). I do think it might make sense to add the ability for the user to disable the object LOD for debugging, though.

Comment (TokisanGames) [reply gdnative/multi-lod/src/lod.cpp:320]: I think this PR is good to go, barring any other testing and changes you wish. These other notes are issues present in the main branch and are an extended scope.

---

RE `enabled`, after looking into how `enabled` and `registered` are used I have a several ideas of how to improve them and save many unnecessary clock cycles during scene changes. I will make a separate PR (branching off of this one) for my proposal and we can talk about it there.

---

RE `set_node_processing`, I'll make another PR and we can discuss it more there. Basically we need a way to disable an entire object when it is out of range, but we don't know the structure of their object. The ImmediateGeometry node is most likely not going to be the root. In mine they are physicsbodies, in your demo they are spatials. We also don't know what the user wants to disable. Currently it just disables the `lod[i]->set_physics/process(disabled)`, which are the 4 mesh instance children that likely do not have any scripts running on them anyway.

A complex object could have multiple scripts running in it's tree. `set_physics_process()` only en/disables the calling of `_physics_process()` for that single node and that's it. Pause mode is inherited when the whole tree is paused, not per object process ticks. The object could have an enemy AI script running on the root physicsbody, a statemachine running on a child node, and animation and sound players, plus physics calculating, none of which will be affected by the call.

I think we should remove `set_node_processing`, which isn't working anyway, and just emit signals as you suggested. That way the user can define their own structure, and they can do whatever they want with the signals. Signals are only sent to connected objects, so there is no danger of sending too many signals, so we don't need an option to enable them. (See Object::emit_signal)

I'm thinking of emitting signals on every LOD change and being unloaded, that way a user can react differently based upon different distances or being killed. Visibility is already signaled by show/hide. Also a user can use `VisibilityNotifier` to en/disable based on frustum and occlusion culling if desired.

As far as letting users know how, I'll make an example in the demo scene and documentation, which should be sufficient.

Comment (puchik) [reply gdnative/multi-lod/src/lod.cpp:320]: Those ideas sound good. The signal alternative makes sense to me. Give me a couple days and this weekend I can try to take a quick test and closer look, and if all is well it'll be good to go. Will also give you more time to make any additional changes. I suspect any changes requested will be minimal, if any, so you can experiment with those other ideas.

**Comment by TokisanGames** [comment]
Just force pushed a change to the 2nd previous commit to remove redundant calls to show_lod, moving `current_lod=lod` up to it's current position.

File: gdnative/multi-lod/src/lod.h [hunk 4]
```diff
@@ -188,18 +191,22 @@ class LOD : public VisualInstance, public LODBaseVariables {
...
 
-    Spatial* lod0 = NULL; // LOD0 MUST exist otherwise screen percentage breaks
-    Spatial* lod1 = NULL; // Plus it doesn't make sense for it not to exist
-    Spatial* lod2 = NULL;
-    Spatial* lod3 = NULL;
+    // LOD0 MUST exist otherwise screen percentage breaks
+    // Plus it doesn't make sense for it not to exist
+    Spatial* lods[LOD_COUNT] = { };
+
+    int last_lod = 0;   // Last active LOD index
```
Comment (puchik) [inline gdnative/multi-lod/src/lod.h]: Can you move this comment to be above the variable and reword it to "The highest level of LOD available/found."

**Comment by puchik** [comment]
One more tiny change to a comment. I'm thinking of rewording the docs a tiny bit but that's fine for now. Also, can you increment the version number to 1.2 for tracking [here](https://github.com/puchik/godot-extras/blob/57e595bfc0a8d787615c03369f1e116bdbbe484d/gdnative/multi-lod/addons/MultiLOD/plugin.cfg#L6)? 

After those two it's good to go!

**Review by puchik — APPROVED** [review]

**Comment by puchik** [comment]
One more tiny change to a comment. I'm thinking of rewording the docs a tiny bit but that's fine for now. Also, can you increment the version number to 1.2 for tracking [here](https://github.com/puchik/godot-extras/blob/57e595bfc0a8d787615c03369f1e116bdbbe484d/gdnative/multi-lod/addons/MultiLOD/plugin.cfg#L6)?

After those two it's good to go!

**Comment by TokisanGames** [comment]
Both done.

**Comment by puchik** [comment]
Thanks!
